src/GCDHelper.cs

```diff
@@ -0,0 +1,616 @@
+using System.Linq;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.RegularExpressions;
+
+
+[assembly: InternalsVisibleTo("Tests")]
+namespace GCDTracker {
+    public class AbilityManager {
+        public record AbilityTiming(float AnimationLock, bool IsCasted);
+        private static AbilityManager instance;
+        public Dictionary<float, AbilityTiming> ogcds { get; private set; }
+        
+        private AbilityManager() {
+            ogcds = [];
+        }
+        public static AbilityManager Instance {
+            get {
+                instance ??= new AbilityManager();
+                return instance;
+            }
+        }
+
+        public void UpdateOGCDs(Dictionary<float, AbilityTiming> newOgcds) {
+            ogcds = newOgcds;
+        }
+    }
+
+    public class BarDecisionHelper {
+        private static BarDecisionHelper instance;
+        public bool Queue_VerticalBar { get; private set; }
+        public bool Queue_Triangle { get; private set; }
+        public bool SlideStart_VerticalBar { get; private set; }
+        public bool SlideEnd_VerticalBar { get; private set; }
+        public bool SlideStart_LeftTri { get; private set; }
+        public bool SlideStart_RightTri { get; private set; }
+        public bool SlideEnd_RightTri { get; private set; }
+        public bool Slide_Background { get; private set; }
+        public float Queue_Lock_Start { get; private set; }
+        public float Slide_Bar_Start { get; private set; }
+        public float Slide_Bar_End { get; private set; }
+
+        private float previousPos = 1f;
+        static readonly float epsilon = 0.02f;
+        
+        private BarDecisionHelper() { }
+        public static BarDecisionHelper Instance {
+            get {
+                instance ??= new BarDecisionHelper();
+                return instance;
+            }
+        }
+        public enum BarState {
+            GCDOnly,
+            ShortCast,
+            LongCast,
+            NonAbilityCast,
+            Mount,
+            Idle
+        }
+        public BarState currentState;
+
+        public void Update(BarInfo bar, Configuration conf, GCDHelper helper) {                
+            if (bar.CurrentPos > (epsilon / bar.TotalBarTime) && bar.CurrentPos < previousPos - epsilon) {
+                // Reset
+                previousPos = 0f;
+                ResetBar(conf);
+
+                // Handle Castbar
+                if(bar.IsCastBar){
+                    Slide_Bar_Start = bar.GCDTime_SlidecastStart;
+                    Slide_Bar_End = conf.SlideCastFullBar ? 1f : bar.GCDTotal_SlidecastEnd;
+                    if (bar.IsNonAbility) {
+                        Queue_Lock_Start = 0f;
+                        Queue_VerticalBar = false;
+                        Queue_Triangle = false;
+                        Slide_Bar_End = 1f;
+                        currentState = helper.queuedAbilityActionType switch
```

Comment:
I was probably overthinking it.  I've changed it for now, but this took a lot of time to validate mostly because I needed to find a place in the MSQ with an event trigger to interact with to test.  From looking at the logs of the currentState and actionType I think we should be okay grabbing it live though.